### PR TITLE
feat(kas):  collect metrics

### PR DESCRIPTION
**Pull request by sujankota**
- Adds [otel](https://opentelemetry.io/) based tracing to kas methods: publicKey and rewrap
  - adds `trace` top level config to viper config, including `trace.enabled` (boolean) for enabling otel tracer, and `trace.folder` for setting the folder to store the `trace.log` file
- Adds new `examples benchmark` command to run encrypt/decrypt and rewrap benchmarks
   - Also adds a job to the checks.yaml to run it, but it is pretty bare-bones for now, and doesn't have any checks against performance regressions or anything.
- Variety of small linter fixes

```
tdf3:
Benchmark Results:
Total Requests: 5000
Successful Requests: 5000
Failed Requests: 0
Concurrent Requests: 50
Total Time: 1m7.704277611s
Average Latency: 673.018399ms
Throughput: 73.85 requests/second

nano:
Benchmark Results:
Total Requests: 5000
Successful Requests: 4992
Failed Requests: 8
Concurrent Requests: 50
Total Time: 57.72426082s
Average Latency: 575.422342ms
Throughput: 86.48 requests/second

Error Summary:
ReadNanoTDF error: readSeeker.Seek failed: error making nano rewrap request to kas: error making rewrap request: rpc error: code = PermissionDenied desc = request error
rpc error: code = PermissionDenied desc = forbidden: 8 occurrences


Trace Statistics:

publickey:
  Total Requests: 2
  Average Duration: 0.073 ms (72.647 µs, 72647 ns)
  Min Duration: 0.030 ms (2[9](https://github.com/opentdf/platform/actions/runs/12053261131/job/33608512613?pr=1702#step:17:10).937 µs, 29937 ns)
  Max Duration: 0.[11](https://github.com/opentdf/platform/actions/runs/12053261131/job/33608512613?pr=1702#step:17:12)5 ms (115.357 µs, 115357 ns)

rewrap-tdf3:
  Total Requests: 5000
  Average Duration: 316.499 ms (316498.907 µs, 316498907 ns)
  Min Duration: 23.608 ms (23607.787 µs, 23607787 ns)
  Max Duration: 979.418 ms (979417.815 µs, 979417815 ns)

rewrap-nanotdf:
  Total Requests: 4776
  Average Duration: 320.718 ms (320717.904 µs, 320717904 ns)
  Min Duration: 22.802 ms (22801.981 µs, 22801981 ns)
  Max Duration: 874.830 ms (874829.7[13](https://github.com/opentdf/platform/actions/runs/12053261131/job/33608512613?pr=1702#step:17:14) µs, 874829713 ns)
```

### CODE REVIEW

**Review by pflynn-virtru — DISMISSED** [review]

File: service/go.mod [hunk 1]
```diff
@@ -15,7 +15,8 @@ require (
 	github.com/go-chi/cors v1.2.1
 	github.com/go-playground/validator/v10 v10.22.0
 	github.com/google/uuid v1.6.0
-	github.com/grpc-ecosystem/grpc-gateway/v2 v2.20.0
+	github.com/grpc-ecosystem/go-grpc-middleware/v2 v2.1.0
```
Comment (dmihalcik-virtru) [inline service/go.mod:18]: Why is this line added?

File: service/kas/access/publicKey.go [hunk 2]
```diff
@@ -11,6 +11,10 @@ import (
 	"connectrpc.com/connect"
 	kaspb "github.com/opentdf/platform/protocol/go/kas"
 	"github.com/opentdf/platform/service/internal/security"
+	"github.com/opentdf/platform/service/tracing"
+	"go.opentelemetry.io/otel"
+	"google.golang.org/grpc/codes"
```
Comment (dmihalcik-virtru) [inline service/kas/access/publicKey.go:16]: Running `make fix` might clean these up, I forget

File: service/main.go [hunk 3]
```diff
@@ -1,7 +1,14 @@
...
 
-import "github.com/opentdf/platform/service/cmd"
+import (
+	"github.com/opentdf/platform/service/cmd"
+	"github.com/opentdf/platform/service/tracing"
+)
 
 func main() {
+	// Initialize tracer
+	shutdown := tracing.InitTracer()
```
Comment (dmihalcik-virtru) [inline service/main.go:10]: Can you file a follow-up issue to make this optional? Even beetter, we would export a tracer API and have applications (e.g. example in this case) inject a Tracer somehow, which would invert the depencency and not require service's go.mod to import otel, etc.

**Review by dmihalcik-virtru — CHANGES_REQUESTED** [review]
Just some nits about cleaning up the imports.

**Review by dmihalcik-virtru — APPROVED** [review]

File: examples/cmd/examples.go [hunk 4]
```diff
@@ -31,7 +31,7 @@ func init() {
 
 func newSDK() (*sdk.SDK, error) {
 	resolver.SetDefaultScheme("passthrough")
-	opts := []sdk.Option{sdk.WithStoreCollectionHeaders()}
```
Comment (dmihalcik-virtru) [inline examples/cmd/examples.go:34]: Is this causing failures in integration test?

**Review by dmihalcik-virtru — DISMISSED** [review]

File: examples/cmd/examples.go [hunk 5]
```diff
@@ -31,7 +31,7 @@ func init() {
...
+	opts := []sdk.Option{sdk.WithStoreCollectionHeaders(), sdk.WithInsecurePlaintextConn()}
```
Comment (imdominicreed) [inline examples/cmd/examples.go:34]: you probably want to drop sdk.WithStoreCollectionHeaders().

There is no way to determine if a nanoTDF is part of a collection or not, with this option on, it will always assume its part of a collection.

Since you are decrypting the same nanoTDF, it will store the header and skip the rewrap call.

It's probably best that you remove this option for benchmarking.

Comment (dmihalcik-virtru) [reply examples/cmd/examples.go]: I've added a parameter, `--storeCollectionHeaders`, to the `examples` command. It defaults to `true`. I've also updated the nano benchmark in CI has it set to `false`

**Comment by jrschumacher** [comment]
@sujankota could we add a [github-script](https://github.com/actions/github-script) step to post the results of the benchmark process to the PR? Something like:

_Credit ChatGPT_
```
- name: Post benchmark results as a PR comment
  if: ${{ github.event_name == 'pull_request' }}
  uses: actions/github-script@v6
  with:
    script: |
      const results = process.env.RESULTS;
      const commentBody = `
<details>
  <summary>Benchmark Results</summary>

  \`\`\`shell
  ${results}
  \`\`\`
</details>
      `;
      const { data: comments } = await github.rest.issues.listComments({
        owner: context.repo.owner,
        repo: context.repo.repo,
        issue_number: context.issue.number,
      });

      const existingComment = comments.find(comment => 
        comment.user.login === 'github-actions[bot]' && 
        comment.body.includes('<summary>Benchmark Results</summary>')
      );

      if (existingComment) {
        await github.rest.issues.updateComment({
          owner: context.repo.owner,
          repo: context.repo.repo,
          comment_id: existingComment.id,
          body: commentBody,
        });
      } else {
        await github.rest.issues.createComment({
          owner: context.repo.owner,
          repo: context.repo.repo,
          issue_number: context.issue.number,
          body: commentBody,
        });
      }
    env:
      RESULTS: ${{ steps.previous_step_name.outputs.results }}
```

File: .github/workflows/checks.yaml [hunk 6]
```diff
@@ -157,6 +157,67 @@ jobs:
...
+          sudo cp ./keys/localhost.crt /usr/local/share/ca-certificates
+          sudo update-ca-certificates
+      - run: docker compose up -d --wait --wait-timeout 240 || (docker compose logs && exit 1)
+      - run: go run ./service provision keycloak
+      - run: go run ./service provision fixtures
+      - uses: JarvusInnovations/background-action@2428e7b970a846423095c79d43f759abf979a635
+        name: start server in background
+        with:
+          run: >
+            go build -o opentdf -v service/main.go
```
Comment (strantalis) [inline .github/workflows/checks.yaml:206]: In a follow up we should leverage the shared workflow for standing up the platform. 

https://github.com/opentdf/platform/blob/main/test/start-up-with-containers/action.yaml

**Review by strantalis — APPROVED** [review]
